allennlp_models/generation/models/t5.py

```diff
@@ -0,0 +1,119 @@
+
+    def forward(  # type: ignore
+        self, source_tokens: TextFieldTensors, target_tokens: Optional[TextFieldTensors] = None
+    ) -> Dict[str, torch.Tensor]:
+        """
+        Performs the forward step of T5.
+
+        # Parameters
+
+        source_tokens : `TextFieldTensors`, required
+            The source tokens for the encoder. We assume they are stored under the `tokens` key/namespace.
+
+        target_tokens : `TextFieldTensors`, optional (default = `None`)
+            The target tokens for the decoder. We assume they are also stored under the `tokens` key/namespace.
+            If no target tokens are given during training / validation, the source tokens are shifted
+            to the right by 1.
+
+        # Returns
+
+        `Dict[str, torch.Tensor]`
+            Contains the `loss` when `target_tokens` is provided.
+            And during prediction, includes `predictions` and `predicted_log_probs` from beam search.
+
+        """
+        input_ids, attention_mask = (
+            source_tokens["tokens"]["token_ids"],
+            source_tokens["tokens"]["mask"],
+        )
+        labels: Optional[IntT] = None
+        decoder_attention_mask: Optional[BoolT] = None
+        if target_tokens is not None:
+            labels, decoder_attention_mask = (
+                target_tokens["tokens"]["token_ids"],  # type: ignore[assignment]
+                target_tokens["tokens"]["mask"],  # type: ignore[assignment]
+            )
+        elif self.training:
+            raise ValueError("'target_tokens' required during training")
+
+        output: T5Output = self.t5(
+            input_ids,
+            attention_mask=attention_mask,
+            labels=labels,
+            decoder_attention_mask=decoder_attention_mask,
+        )
+        output_dict: Dict[str, torch.Tensor] = {}
+
+        if self.training:
+            assert output.loss is not None
+            output_dict["loss"] = output.loss
+        else:
+            # Shape: (batch_size, beam_size, num_tokens)
+            assert output.predictions is not None
+            # Shape: (batch_size, beam_size)
+            assert output.predicted_log_probs is not None
+            # Shape: (batch_size, num_tokens)
+            output_dict["predictions"] = output.predictions[:, 0, :]
+            # Shape: (batch_size, )
+            output_dict["predicted_log_probs"] = output.predicted_log_probs[:, 0]
+
+            if labels is not None:
+                assert output.loss is not None
+                output_dict["loss"] = output.loss
+
+                for metric in self._metrics:
+                    metric(output_dict["predictions"], labels)  # type: ignore[call-arg]
+
+        return output_dict
+
+    @overrides
+    def make_output_human_readable(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, Any]:
+        predictions = output_dict["predictions"]
+        output_dict["predicted_text"] = self.tokenizer.tokenizer.batch_decode(
```

Comment:
The normal way to do this is by looking into the vocab. Then you don't need a tokenizer in the model at all.

Reply:
This does more than just convert IDs into tokens. It also de-tokenizes the tokens while removing any special tokens. Sure, we could re-write that logic here using just the vocab, but I don't think it's worth it.

Reply:
Fair enough